test/extended/oauth/oauthcertfallback.go

```diff
@@ -0,0 +1,151 @@
+package oauth
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/rest"
+	restclient "k8s.io/client-go/rest"
+
+	"github.com/openshift/origin/pkg/cmd/server/admin"
+	userclient "github.com/openshift/origin/pkg/user/generated/internalclientset/typed/user/internalversion"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+const (
+	invalidToken      = "invalid"
+	noToken           = ""
+	certUser          = "system:admin"
+	unauthorizedError = "Unauthorized"
+	anonymousError    = `users.user.openshift.io "~" is forbidden: User "system:anonymous" cannot get resource "users" in API group "user.openshift.io" at the cluster scope`
+)
+
+var (
+	invalidCert = restclient.TLSClientConfig{}
+	noCert      = restclient.TLSClientConfig{}
+)
+
+var _ = g.Describe("The OAuth server", func() {
+	defer g.GinkgoRecover()
+	oc := exutil.NewCLI("oauth", exutil.KubeConfigPath())
+
+	g.It("has the correct token and certificate fallback semantics", func() {
+		fakecadir, err := ioutil.TempDir("", "fakeca")
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		defer os.RemoveAll(fakecadir)
+
+		fakeca, err := (&admin.CreateSignerCertOptions{
+			CertFile:   path.Join(fakecadir, "fakeca.crt"),
+			KeyFile:    path.Join(fakecadir, "fakeca.key"),
+			SerialFile: path.Join(fakecadir, "fakeca.serial"),
+			Name:       "foo",
+			IOStreams:  genericclioptions.NewTestIOStreamsDiscard(),
+			Overwrite:  true,
+		}).CreateSignerCert()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		clientCertConfig, err := fakeca.MakeClientCertificate(
+			path.Join(fakecadir, "fakeclient.crt"),
+			path.Join(fakecadir, "fakeclient.key"),
+			&user.DefaultInfo{Name: "fakeuser"},
+			365*2, /* 2 years */
+		)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		invalidCert.CertData, invalidCert.KeyData, err = clientCertConfig.GetPEMBytes()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		tokenUser := fmt.Sprintf("system:serviceaccount:%s:builder", oc.Namespace())
+
+		validToken, err := oc.Run("sa").Args("get-token", "builder").Output()
```

Comment:
You need a token for an actual user otherwise the semantics of the e2e do not line up with the integration test, see `GetClientForUser`.